language/bytecode-verifier/src/dependencies.rs

```diff
@@ -338,3 +338,46 @@ impl<'a> DependencyChecker<'a> {
         }
     }
 }
+
+pub struct CyclicModuleDependencyChecker {}
+
+impl CyclicModuleDependencyChecker {
+    pub fn verify_module<F>(module: &CompiledModule, dep_loader: F) -> VMResult<()>
+    where
+        F: Fn(&ModuleId) -> Vec<ModuleId>,
+    {
+        Self::verify_module_impl(module, dep_loader)
+            .map_err(|e| e.finish(Location::Module(module.self_id())))
+    }
+
+    fn verify_module_impl<F>(module: &CompiledModule, dep_loader: F) -> PartialVMResult<()>
+    where
+        F: Fn(&ModuleId) -> Vec<ModuleId>,
+    {
+        // A standard DFS algorithm that follows the dependency chain from the `cursor_module_id`
+        // to see whether the `target_module_id` is in its dependencies.
+        fn is_reachable_via_dependencies<F>(
+            target_module_id: &ModuleId,
+            cursor_module_id: &ModuleId,
+            dep_loader: &F,
+        ) -> bool
+        where
+            F: Fn(&ModuleId) -> Vec<ModuleId>,
+        {
+            target_module_id == cursor_module_id
+                || (dep_loader)(cursor_module_id)
```

Comment:
Do you need the parens still?

Reply:
Technically no, it's more like a personal preference to indicate that this is a function pointer...

Reply:
They aren't function pointers though... at least not really. I mean at some point sure they are, but as far as Rust is concerned you have some value that implements the trait `Fn(&ModuleId) -> Vec<ModuleId>`

I think the consistent style with the rest of the codebase would be to not include them